Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -413,6 +433,128 @@ DocumentKey Serializer::DecodeKey(absl::string_view name) const {
+  // Encode Document.name
+  writer->WriteTag({PB_WT_STRING, google_firestore_v1beta1_Document_name_tag});
+  writer->WriteString(EncodeKey(key));
+
+  // Encode Document.fields (unless it's empty)
+  if (!object_value.internal_value.empty()) {
+    writer->WriteTag(
+        {PB_WT_STRING, google_firestore_v1beta1_Document_fields_tag});
+    EncodeObject(writer, object_value);
+  }
+
+  // Skip Document.create_time and Document.update_time, since they're output
+  // only fields.
+}
+
+util::StatusOr<std::unique_ptr<model::MaybeDocument>>
+Serializer::DecodeMaybeDocument(const uint8_t* bytes, size_t length) const {
+  Reader reader = Reader::Wrap(bytes, length);
+  std::unique_ptr<MaybeDocument> maybeDoc =
+      DecodeBatchGetDocumentsResponse(&reader);
+
+  if (reader.status().ok()) {
+    return maybeDoc;
+  } else {
+    return reader.status();
+  }
+}
+
+std::unique_ptr<MaybeDocument> Serializer::DecodeBatchGetDocumentsResponse(
+    Reader* reader) const {
+  Tag tag = reader->ReadTag();
+  if (!reader->status().ok()) return InvalidMaybeDocument();
+
+  // Ensure the tag matches the wire type
+  switch (tag.field_number) {
+    case google_firestore_v1beta1_BatchGetDocumentsResponse_found_tag:
+    case google_firestore_v1beta1_BatchGetDocumentsResponse_missing_tag:
+      if (tag.wire_type != PB_WT_STRING) {
+        reader->set_status(
+            Status(FirestoreErrorCode::DataLoss,
+                   "Input proto bytes cannot be parsed (mismatch between "
+                   "the wiretype and the field number (tag))"));
+      }
+      break;
+
+    default:
+      reader->set_status(Status(
+          FirestoreErrorCode::DataLoss,
+          "Input proto bytes cannot be parsed (invalid field number (tag))"));
+  }
+
+  if (!reader->status().ok()) return InvalidMaybeDocument();
+
+  switch (tag.field_number) {
+    case google_firestore_v1beta1_BatchGetDocumentsResponse_found_tag:
+      return reader->ReadNestedMessage<std::unique_ptr<MaybeDocument>>(
+          InvalidMaybeDocument(),
+          [this](Reader* reader) -> std::unique_ptr<MaybeDocument> {
+            return DecodeDocument(reader);
+          });
+    case google_firestore_v1beta1_BatchGetDocumentsResponse_missing_tag:
+      // TODO(rsgowman): Right now, we only support Document (and don't support
+      // NoDocument). That should change in the next PR or so.
+      abort();
+    default:
+      // This indicates an internal error as we've already ensured that this is
+      // a valid field_number.
+      FIREBASE_ASSERT_MESSAGE(
+          false,
+          "Somehow got an unexpected field number (tag) after verifying that "
+          "the field number was expected.");
+  }
+}
+
+std::unique_ptr<Document> Serializer::DecodeDocument(Reader* reader) const {
+  std::string name = "";
+  FieldValue fields = FieldValue::ObjectValueFromMap({});
+  SnapshotVersion version = SnapshotVersion::None();
+
+  while (reader->bytes_left()) {
```

Comment:
Is the loop necessary? If I read it correctly, it means that, if the function were given bytes that contain "fields" twice (which, IIUC, is invalid), it would accept the input, read the first "fields" and then overwrite them with second "fields". Is it possible/valid for any of the tags to be repeated? Can the ordering of tags vary?

Reply:
This is the only substantial comment, otherwise pretty much LGTM.

Reply:
Thanks for explaining! Sorry for nagging on this, but consider adding a very brief explanation to the comments, perhaps just a shortened quote from the docs:
```
// https://developers.google.com/protocol-buffers/docs/encoding:
// "Normally, an encoded message would never have more than one instance of a non-repeated field.
// However, parsers are expected to handle the case in which they do. For numeric types and strings,
// if the same field appears multiple times, the parser accepts the last value it sees. For embedded 
// message fields, the parser merges multiple instances of the same field."
```
Optional, but I think it could be helpful.